FLT/NumberField/AdeleRing.lean

```diff
@@ -18,6 +23,141 @@ end LocallyCompact
+  LinearEquiv.trans (TensorProduct.congr (LinearEquiv.refl K (AdeleRing K))
+      (Basis.equivFun (Module.finBasis K L)))
+    (TensorProduct.piScalarRight _ _ _ _)
+
+variable {L}
+
+theorem tensorProductLinearEquivPi_tsum_apply (l : L) :
+    tensorProductLinearEquivPi K L (1 ⊗ₜ[K] l) =
+      fun i => algebraMap _ _ (Module.finBasis K L |>.equivFun l i) := by
+  simp [tensorProductLinearEquivPi, Algebra.algebraMap_eq_smul_one]
+
+variable {K}
+
+theorem tensorProductLinearEquivPi_symm_apply_of_algebraMap
+    (x : Fin (Module.finrank K L) → K) :
+    (tensorProductLinearEquivPi K L).symm (fun i => algebraMap _ _ (x i)) =
+      1 ⊗ₜ[K] ((Module.finBasis K L).equivFun.symm x) := by
+  simp only [LinearEquiv.trans_symm, LinearEquiv.trans_apply,
+    Algebra.TensorProduct.piScalarRight_symm_apply_of_algebraMap, TensorProduct.congr_symm_tmul,
+    LinearEquiv.refl_symm, LinearEquiv.refl_apply, map_sum, Basis.equivFun_symm_apply]
+  rw [Finset.sum_comm]
+  simp only [← Finset.sum_smul, Fintype.sum_pi_single]
+
+/-
+smmercuri: A possible alternative using `moduleTopology` is
+```
+instance :
+    TopologicalSpace (NumberField.AdeleRing K ⊗[K] L) :=
+  letI := TopologicalSpace.induced (algebraMap K (AdeleRing K)) inferInstance
+  moduleTopology K _
+```
+However it is not clear to me how the inverse function is of `tensorProductLinearEquivPi`
+is continuous in that case. Additionally,
+https://math.mit.edu/classes/18.785/2017fa/LectureNotes25.pdf (just above Prop 25.10)
+for an informal source where the tensor product is given the product topology. Maybe they
+coincide!
+-/
+instance : TopologicalSpace (NumberField.AdeleRing K ⊗[K] L) :=
+  TopologicalSpace.induced (NumberField.AdeleRing.tensorProductLinearEquivPi K L) inferInstance
+
+variable (K L)
+
+noncomputable def tensorProductContinuousLinearEquivPi :
+    AdeleRing K ⊗[K] L ≃L[K] (Fin (Module.finrank K L) → AdeleRing K) where
+  toLinearEquiv := tensorProductLinearEquivPi K L
+  continuous_toFun := continuous_induced_dom
+  continuous_invFun := by
+    convert (tensorProductLinearEquivPi K L).toEquiv.coinduced_symm ▸ continuous_coinduced_rng
+
+variable {K L}
+
+theorem tensorProductContinuousLinearEquivPi_symm_apply_of_algebraMap
+    (x : Fin (Module.finrank K L) → K) :
+    (tensorProductContinuousLinearEquivPi K L).symm (fun i => algebraMap _ _ (x i)) =
+      1 ⊗ₜ[K] ((Module.finBasis K L).equivFun.symm x) := by
+  exact tensorProductLinearEquivPi_symm_apply_of_algebraMap _
+
+variable (K L)
+
+/-
+smmercuri : The tensor product here is in a different order to the one
+appearing in the finite adele ring base change, where `L ⊗[K] FiniteAdeleRing A K`
+is used. Is there a preference between these orderings? One benefit
+to using `AdeleRing K ⊗[K] L` is that it automatically gets
+a `K` algebra instance via the instance `Algebra.TensorProduct.leftAlgebra`
+(while `rightAlgebra` is a def), but maybe there are other reasons to
+prefer the `rightAlgebra` set up.
```

Comment:
Yeah :-/ Of course the benefit of using the other order is that you automatically get an `L`-algebra instance. I am still really unclear about the best way to proceed here. When doing automorphic forms it's really normal to have G(Q) \ G(A) and so I've been putting number fields on the left in general but that was my only logic.